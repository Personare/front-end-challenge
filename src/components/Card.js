import React, { PropTypes } from 'react';
import TweenLite from 'gsap';

export class Card extends React.Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.selected !== this.props.selected) {
            TweenLite.to(this.refs.cardFaces, 1, { rotationY: !nextProps.selected ? -180 : 0 });
        }
    }

    render() {
        return (
            <article className="card" onClick={this.props.onClick}>
                <div className="card-faces" ref="cardFaces">
                    <div className="front">
                        <img src={this.props.frontImg} alt={this.props.name} />
                    </div>
                    <div className="back">
                        <img src={this.props.backImg} alt={this.props.name} />
                    </div>
                </div>
            </article>
        );
    }
}

Card.propTypes = {
    name: PropTypes.string,
    frontImg: PropTypes.string,
    backImg: PropTypes.string,
    selected: PropTypes.bool,
    onClick: PropTypes.func,
};
