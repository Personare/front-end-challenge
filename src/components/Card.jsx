import React from "react";

const Card = ({ name = "CartjÃªnha", image = "", ...rest }) => {
  return (
    <div className="card">
      <img src={image} alt={name} />
    </div>
  );
};

export default Card;
