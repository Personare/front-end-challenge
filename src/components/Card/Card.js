import React from 'react'
import styled from 'styled-components'
import { string, func, bool } from 'prop-types'

const Card = ({
  name,
  description,
  image,
  imagesUrl,
  imageBackCard,
  hasStart,
  onClick,
  width,
  height,
  cardBoard
}) => (
  <Wrapper onClick={onClick} image={image} cardBoard={cardBoard}>
    {name && <CardName>{name}</CardName>}
    <ImgWrapper
      image={image}
      imagesUrl={imagesUrl}
      imageBackCard={imageBackCard}
      hasStart={hasStart}
      width={width}
      height={height}
    />
    {description && <CardDescription>{description}</CardDescription>}
  </Wrapper>
)

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 10px;
  box-shadow: ${({ cardBoard }) => (cardBoard ? '5px 5px 5px 0px rgba(0, 0, 0, 0.75)' : 'none')};
`

const CardName = styled.div`
  margin-bottom: 20px;
`
const CardDescription = styled.div`
  max-width: 300px;
  text-align: justify;
  margin-top: 20px;
`

const ImgWrapper = styled.img.attrs({
  src: ({
    image, imagesUrl, imageBackCard, hasStart
  }) => (hasStart ? imageBackCard : `${imagesUrl}${image}`)
})`
  width: ${({ width }) => width};
  height: ${({ height }) => height};
  background: transparent url('./assets/loading.gif') center no-repeat;
`

Card.propTypes = {
  name: string,
  description: string,
  image: string.isRequired,
  imagesUrl: string.isRequired,
  imageBackCard: string.isRequired,
  hasStart: bool,
  onClick: func,
  width: string,
  height: string,
  cardBoard: bool
}

Card.defaultProps = {
  name: '',
  description: '',
  hasStart: false,
  width: '100px',
  height: '200px',
  cardBoard: true,
  onClick: Function
}

export default Card
