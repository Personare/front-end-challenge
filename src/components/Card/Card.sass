.CardContainer
  position: relative
  height: 210px
  cursor: pointer
  margin: 10px;
  .CardContent
    transform-style: preserve-3d

  .CardImage, .BackImage
    position: absolute
    width: 100px
    height: 210px
    overflow: hidden
    border-radius: 5px
    backface-visibility: hidden
    > img
      width: 100px
      height: 210px

  .BackImage
    transform: rotateY(-180deg)
