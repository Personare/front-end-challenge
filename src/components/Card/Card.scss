.Card {
    position: relative;
    perspective: 150rem;
    width: 200px;
    height: 200px;

    &__side {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 50px;
        backface-visibility: hidden;
        transition: all .8s ease;

        &.--front,
        &.--back {
            svg {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                left: 0;
                width: 1em;
                height: 1em;
                font-size: 5.4rem;
            }
        }

        &.--front {
            background-color: red;

            svg {
                fill: black;
            }
        }

        &.--back {
            background-color: blue;
            transform: rotateY(180deg);

            svg {
                fill: red;
            }
        }
    }

    &.--closed &__side.\--front {
        transform: rotateY(-180deg);
    }

    &.--closed &__side.\--back {
        transform: rotateY(0);
    }
}

