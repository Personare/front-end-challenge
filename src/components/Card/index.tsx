import { useContext } from "react";
import ReactCardFlip from "react-card-flip";
import { ICard } from "../../config/interfaces";
import GlobalContext from "../../global/GlobalContext";

interface ICardItem {
  card: ICard;
  onClick?: (name: string) => void;
}

const Card: React.FC<ICardItem> = ({ card, onClick }) => {
  const { name, image, show } = card;

  const {
    state: { data },
  }: any = useContext(GlobalContext);

  const isFlipped = !show;

  console.log(onClick);

  return (
    <div data-cy="card" onClick={() => onClick && onClick(name)}>
      <ReactCardFlip isFlipped={isFlipped} flipDirection="horizontal">
        <img
          alt={name}
          src={`${data.imagesUrl}${image}`}
          data-cy={!isFlipped && "card-up"}
        />

        <img
          alt={name}
          src={data.imageBackCard}
          data-cy={isFlipped && "card-down"}
        />
      </ReactCardFlip>
    </div>
  );
};

export default Card;
