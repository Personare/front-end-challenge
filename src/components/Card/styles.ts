import styled, { css } from "styled-components";

interface ImageProps {
  isMinimal?: boolean;
}
export const Image = styled.img<ImageProps>`
  ${({ isMinimal }) => css`
    height: ${isMinimal ? "150px" : "300px"};
    width: ${isMinimal ? "79px" : "auto"};
    margin-bottom: ${isMinimal ? "10px" : "auto"};
  `}
`;

export const Container = styled.div``;
