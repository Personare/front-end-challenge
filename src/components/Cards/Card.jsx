import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import styles from './Card.module.css'
import CardName from './CardName'
import CardImage from '../../containers/Cards/CardImage'

class Card extends Component {
  handleCardClick = () => {
    const { isShuffle } = this.props

    if (isShuffle) {
      this.setValidCurrentCard()
    }
  }

  setValidCurrentCard() {
    const { currentCard, setCurrentCard, item } = this.props

    if (!currentCard.id) {
      setCurrentCard(item)
    }
  }

  render() {
    const { isShuffle, item } = this.props

    return (
      <div
        className={classNames(styles.card, {
          [styles.backcard]: isShuffle,
        })}
        onClick={this.handleCardClick}
      >
        <CardImage isShuffle={isShuffle} item={item} />
        <CardName isShuffle={isShuffle} text={item.name} />
      </div>
    )
  }
}

Card.propTypes = {
  isShuffle: PropTypes.bool.isRequired,
  setCurrentCard: PropTypes.func.isRequired,
  item: PropTypes.shape({
    name: PropTypes.string.isRequired,
    imageUrl: PropTypes.string.isRequired,
  }).isRequired,
}

export default Card
