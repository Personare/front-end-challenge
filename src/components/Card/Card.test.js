import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import Card from './Card'

describe('Card component', () => {
  const props = {
    name: 'O MAGO',
    image: 'arcano1.jpg',
    imagesUrl: `https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/
        images/content/pt-br/product/tarot/marselha/162x341/`,
    imageBackCard: `https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/
        images/content/pt-br/product/tarot/marselha/back-blue-card.png`,
    hasStart: false,
    onClick: Function,
    flipCard: false
  }

  it('should render correctly', () => {
    const tree = renderer.create(<Card {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should show the front of the card when game is not start', () => {
    const { image, imagesUrl } = props
    const tree = mount(<Card {...props} />)
    const imgSrc = `${imagesUrl}${image}`
    const imgWrapper = tree.find('img')

    expect(imgWrapper.prop('src')).toBe(imgSrc)
  })

  it('should show the back of the card when game was started', () => {
    const { imageBackCard } = props

    const tree = mount(<Card {...props} hasStart />)
    const imgWrapper = tree.find('img')

    expect(imgWrapper.prop('src')).toBe(imageBackCard)
  })
})
