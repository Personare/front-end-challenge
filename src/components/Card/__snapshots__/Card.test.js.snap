// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render correctly 1`] = `
<div
  className="sc-bdVaJa lhHYPK"
  onClick={[Function]}
>
  <div
    className="sc-bwzfXH gHzNhV"
  >
    O MAGO
  </div>
  <img
    className="sc-bxivhb eIjJhB"
    height="200px"
    src="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/
        images/content/pt-br/product/tarot/marselha/162x341/arcano1.jpg"
    width="100px"
  />
  
</div>
`;
